File: palindrome_linked_list.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *next = nullptr, *prev = nullptr;
        
        while(head) {
            next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
    bool isPalindrome(ListNode* head) {
        ListNode *slow = head, *fast = head;

        while(fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* tail = reverseList(slow);

        while(head && tail) {
            if(head->val != tail->val) {
                return false;
            }
            head = head->next;
            tail = tail->next;
        }
        return true;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(1)

File: palindrome_linked_list.go
func reverseList(head *ListNode) *ListNode {
    var next, prev *ListNode
    
    for head != nil {
        next = head.Next
        head.Next = prev
        prev = head
        head = next
    }
    return prev
}

func isPalindrome(head *ListNode) bool {
    slow, fast := head, head

    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }
    tail := reverseList(slow)

    for head != nil && tail != nil {
        if head.Val != tail.Val {
            return false
        }
        head = head.Next
        tail = tail.Next
    }
    return true
}

// Time Complexity : O(n)
// Space Complexity : O(1)